### PR TITLE
Extend error handling to all API calls and mock errors

### PR DESCRIPTION
**Pull request by lmangall**
In this PR the debugging is extended to the rest of the API calls, and the debugging from the Resolvers (API mock) is triggered by a new env var

- [x] move CustomError to types file
- [x] add error handling to get-user
- [x] add error handling to add-user-to-agreed
- [x] add error handling to send-code
- [x] add error handling to send-message
- [x]  add .env triggered debugging to each resolver

To add to the .env
VITE_DEBUG_ENDPOINT=add-user-to-agreed  # Can be get-user, add-user-to-agreed, send-code, login, send-message or false
`

### CODE REVIEW

**Comment by 552020** [comment]
I'm checking this PR and trying to understand it. I started to check `/get-user`. You are not handling the error we get from the real API call (in utils.ts getUserDataFromBackend) but in the mocked function `getUserResolver`. If the env variable VITE_DEBUG_ENDPOINT  is set to "get-user", it will return a 500. @lmangall

**Comment by lmangall** [comment]
@552020  You are right, I forgot to do it in utils.ts for get-user (I did it in ConfirmSale.tsx)
I added it now

**Comment by lmangall** [comment]
>  If the env variable VITE_DEBUG_ENDPOINT is set to "get-user", it will return a 500

Yes this is expected


right now there is a problem with the get-user, the error is printed in console.log but not shown in the UI

**Comment by lmangall** [comment]
It's done now, thanks for spotting the missing one

**Comment by 552020** [comment]
But /get-user can return: 400 (if the user is missing from the request), 404 (if the user is missing from the database) and 500. 
- 404 is the normal case for a new user or a browser user: so we should do nothing probably.
- 400 will probably never be the case, cause we send a default value. But we should catch the error anyway maybe.
- 500: how do you want to handle it? We are not able to retrieve the info of the user, and we tell the user what?

File: src/utils/utils.ts [hunk 1]
```diff
@@ -39,24 +39,22 @@ export const getUserDataFromBackend = async (
...
+      const error: CustomError = new Error(errorMessage);
+      error.status = response.status;
+      throw error;
     }
+
     const data = await response.json();
     console.log("User data from backend:", data);
     return data;
   } catch (error) {
-    console.error("Error fetching user data:", error);
```
Comment (552020) [inline src/utils/utils.ts:49]: This is not good I think. 400 is the normal case for the new user.

**Review by 552020 — CHANGES_REQUESTED** [review]

**Comment by lmangall** [comment]
> But /get-user can return: 400 (if the user is missing from the request), 404 (if the user is missing from the database) and 500.


Yes I know but I don't want to reimplement the whole backend to test each code, if one code works all the other should work. Eventually we can do it later, but now the goal is to have the minimal thing, that as instance would show to a user that we don't support 2FA.
The different actions to be taken depending on which error come from which endpoint will be handled in BackendError I think

**Comment by 552020** [comment]
> > But /get-user can return: 400 (if the user is missing from the request), 404 (if the user is missing from the database) and 500.
> 
> Yes I know but I don't want to reimplement the whole backend to test each code, if one code works all the other should work. Eventually we can do it later, but now the goal is to have the minimal thing, that as instance would show to a user that we don't support 2FA. The different actions to be taken depending on which error come from which endpoint will be handled in BackendError I think

What do you mean. I don't understand your message. Why should we reimplement the whole backend to test each code? These are the codes we get now. I was talk exclusively about /get-user now. 

If you need someone who just approve your request maybe you can ask @Danielg1406. He should be at school now.

**Comment by 552020** [comment]
The branch has also conflicts.

**Comment by lmangall** [comment]
> What do you mean? I don't understand your message. Why should we reimplement the whole backend to test each code? These are the codes we get now. I was talk exclusively about /get-user in my message.

I thought you were talking about the mocks and meant that we have to enhance the mocks with more errors

Conflicts fixed


> If you need someone who just approve your request maybe you can ask @Danielg1406. He should be at school now.

Yes he is here, I appreciate your error checking and the fact that you found an API call unhandled !

**Comment by 552020** [comment]
image

So you didn't change even this. @lmangall 

Happy you find someone who approve whatever you write! Great job! 👏

**Comment by lmangall** [comment]
> So you didn't change even this. @lmangall


@552020  I asked you for clarification

Screenshot 2024-07-06 at 14 22 57

**Comment by 552020** [comment]
> > So you didn't change even this. @lmangall
> 
> @552020 I asked you for clarification
> 
> Screenshot 2024-07-06 at 14 22 57

And you waited for the clarification I see. What there is to clarify? For a new user which is not in the db the backend will return a 400. You can just read which errors the backend returns (and in which cases) since you are implementing errors (or pretending to).

File: src/mocks/resolvers/sendMessageResolver.ts [hunk 2]
```diff
@@ -7,53 +8,65 @@ interface SendMessageRequestBody {
...
+      error.status = 400;
+      throw error;
+    }
 
-  const body = json as SendMessageRequestBody;
-  const {chats} = body;
+    if (import.meta.env.VITE_DEBUG_ENDPOINT === "send-message") {
+      const error: CustomError = new Error("Debugging mode: Forced error");
+      error.status = 500;
+      throw error;
```
Comment (552020) [inline src/mocks/resolvers/sendMessageResolver.ts:32]: error.status = import.meta.env.VITE_DEBUG_ENDPOINT_ERROR

**Review by 552020 — CHANGES_REQUESTED** [review]
Please be more specific with the error

**Review by 552020 — APPROVED** [review]
LGTM!

**Pull request merged**